service/storage_proxy.cc

```diff
@@ -2517,18 +2517,33 @@ storage_proxy::mutate_atomically(std::vector<mutation> mutations, db::consistenc
             m.partition().apply_delete(*schema, clustering_key_prefix::make_empty(), tombstone(now, gc_clock::now()));
 
             tracing::trace(_trace_state, "Sending a batchlog remove mutation");
-            return send_batchlog_mutation(std::move(m), db::consistency_level::ANY).handle_exception([] (std::exception_ptr eptr) {
-                slogger.error("Failed to remove mutations from batchlog: {}", eptr);
+            return send_batchlog_mutation(std::move(m), db::consistency_level::ANY).then_wrapped([] (future<result<>> f) {
+                auto print_exception = [] (const auto& ex) {
+                    slogger.error("Failed to remove mutations from batchlog: {}", ex);
+                };
+                if (f.failed()) {
+                    print_exception(f.get_exception());
+                } else if (result<> res = f.get(); res.has_error()) {
+                    res.assume_error().accept(print_exception);
+                }
             });
         };
 
-        future<> run() {
+        future<result<>> run() {
             return _p.mutate_prepare(_mutations, _cl, db::write_type::BATCH, _trace_state, _permit).then([this] (unique_response_handler_vector ids) {
-                return sync_write_to_batchlog().then([this, ids = std::move(ids)] () mutable {
+                return sync_write_to_batchlog().then([this, ids = std::move(ids)] (result<> res) mutable {
+                    if (res.has_error()) {
```

Comment:
Yeah, all this ifs is not a nice part of using the `outcome` :(

Reply:
The `outcome` library defines macros which help reduce this kind of boilerplate ([link](https://www.boost.org/doc/libs/develop/libs/outcome/doc/html/tutorial/essential/result/try.html) to the relevant tutorial section). Maybe we could use them, and also define our versions which work inside functions that return futures? I don't usually like using macros, but perhaps the reduction in boilerplace is worth it.

There are two flavors, the first one is ugly but the second one requires a GCC/Clang extension (which, I guess is OK because we do not use other compilers):

```c++
BOOST_OUTCOME_TRY (auto i, BigInt::fromString(text));
// vs.
int i = BOOST_OUTCOME_TRYX (BigInt::fromString(text)); // It's an expression, so more flexible
```

Reply:
Well, we do not like macros here, but I think meta progarmming is strong enough to provide something like:

```
template <typename L>
auto wrap(L lambda) {
   using ret_type = return_type<L>::value;
   using param_type = param_type<L>::value;

   return [lambda] (param_type v) -> ret_type {
        if (!v) {
            return v;
        }
        return lambda(v);
   }; 
}
```

Reply:
I agree about disliking macros, I prefer a little extra boilerplate to them. Also agree a little metaprogramming (well-protected with concepts) can help.

Reply:
Done, added `utils::result_wrap`.